DemoApp/Sources/Components/SessionTimer/SessionTimer.swift

```diff
@@ -0,0 +1,141 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Combine
+import StreamVideo
+import StreamVideoSwiftUI
+import SwiftUI
+
+@MainActor class SessionTimer: ObservableObject {
+    
+    @Published var showTimerAlert: Bool = false {
+        didSet {
+            if showTimerAlert, let timerEndsAt {
+                sessionEndCountdown?.invalidate()
+                secondsUntilEnd = timerEndsAt.timeIntervalSinceNow
+                sessionEndCountdown = Timer.scheduledTimer(
+                    withTimeInterval: 1.0,
+                    repeats: true,
+                    block: { [weak self] _ in
+                        guard let self else { return }
+                        Task { @MainActor in
+                            if self.secondsUntilEnd <= 0 {
+                                self.sessionEndCountdown?.invalidate()
+                                self.sessionEndCountdown = nil
+                                self.secondsUntilEnd = 0
+                                self.showTimerAlert = false
+                                return
+                            }
+                            self.secondsUntilEnd -= 1
+                        }
+                    }
+                )
+            } else if !showTimerAlert {
+                sessionEndCountdown?.invalidate()
+                secondsUntilEnd = 0
+            }
+        }
+    }
+    
+    @Published var secondsUntilEnd: TimeInterval = 0
+
+    private var call: Call?
+    private var cancellables = Set<AnyCancellable>()
+    private var timerEndsAt: Date? {
+        didSet {
+            setupTimerIfNeeded()
+        }
+    }
+
+    private var timer: Timer?
+    private var sessionEndCountdown: Timer?
+    
+    private let alertInterval: TimeInterval
+    
+    private var extendDuration: TimeInterval
+    
+    private let changeMaxDurationPermission = Permission(
+        rawValue: OwnCapability.changeMaxDuration.rawValue
+    )
+    
+    let extensionTime: TimeInterval
+    
+    var showExtendCallDurationButton: Bool {
+        call?.state.ownCapabilities.contains(.changeMaxDuration) == true
+    }
+
+    @MainActor init(
+        call: Call?,
```

Comment:
If `call` is nil, is there a point to init this object at all? Seems like the call object is required.

Reply:
it's optional from the view model, so best to keep it like this.